### PR TITLE
Blog Post - Improving Stream Data Quality With Protobuf Schema Validation

### PR DESCRIPTION
**Pull request by tomseddon**

### CODE REVIEW

File: _posts/2019-01-10-improving-stream-data-quality-with-protobuf-schema-validation.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+Both of these requirements meant that a flexible data format such as JSON would not be ideal and the team invested in researching transport encoding formats the would provide stronger assurances about the data being transmitted. While there are some ways to give greater guarantees for JSON, using for example JSON Schema, this still leaves a lot to be desired, including a lack of well defined mechanisms for schema evolution, and still leaves us with sub-par encoding and decoding performance of JSON itself.
+
+## Deciding on an Encoding Format
+The team began investigating the range of technologies available for inter-operable schema formats. We were looking for a system that supports defining a schema in a programming language agnostic Interface Definition Language (IDL) which could then propagate the schema across to all the applications that need to work on that data. In addition to this, benefits such as binary serialization (reduced payload size) and schema evolution mechanisms were aspects the team had worked with before on previous projects, and were keen to make use of again.
```
Comment (shs96c) [inline _posts/2019-01-10-improving-stream-data-quality-with-protobuf-schema-validation.md]: But why an IDL? Might be worth pointing out that we use a ton of languages here.

File: _posts/2019-01-10-improving-stream-data-quality-with-protobuf-schema-validation.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+The options available for these types of serialization formats supporting cross language support were narrowed down to three that broadly met our requirements. These were Thrift, Protocol Buffers (Protobuf), and Avro. We then proceeded to conduct an evaluation of these formats to determine what would work best for transmission of data over Kafka.
```
Comment (shs96c) [inline _posts/2019-01-10-improving-stream-data-quality-with-protobuf-schema-validation.md]: You can make this shorter and punchier: "We quickly narrowed the choice of serialisation formats to three: Thrift, protobug, and avro"

Also, you appear to have slipped to a US spelling for "serialisation"

File: _posts/2019-01-10-improving-stream-data-quality-with-protobuf-schema-validation.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+Thrift and Protobuf have very similar semantics, with IDLs that support the broad types and data structures utilised in mainstream programming languages. When conducting our evaluation, we initially chose Thrift due to familiarity, but in the end discounted this due to lack of community support. Avro was an intriguing option, particularly because of Confluent’s support for this on Kafka. Avro semantics are quite different to that of Protobuf, as it is typically used with a schema definition provided in a header to a file. Confluent’s schema registry removes this requirement by keeping the schema definition in an API and tagging each message with a lookup to find that schema. One of the other appealing aspects of Avro is that it manages schema evolution and backwards and forwards compatibility for you, by keeping track of a writers and a readers schema.
```
Comment (shs96c) [inline _posts/2019-01-10-improving-stream-data-quality-with-protobuf-schema-validation.md]: Someone may mention fbthrift.

File: _posts/2019-01-10-improving-stream-data-quality-with-protobuf-schema-validation.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+### Backwards Compatibility
+Backwards compatibility means that consumers using a newer version of the schema can read the data produced by a client with an earlier version of the schema. Fields that have been added in the newer version will again be deserialized with default values. Fields that have been deleted in the new schema will not be deserialized, but naturally any subsequent code that was in place to handle that data will need to be refactored to cope.
+
+The remaining Protobuf requirements that are mandated to ensure data consistency are met by ensuring that the ordinal placeholders for each attribute are held immutable throughout a message’s lifespan. In Protobuf the integer field number that represents a field is sacred, as this is what is actually transmitted in a serialized message (as opposed to the field name). All producers and consumers rely on this integer having a consistent meaning, and altering it can cause havoc if a consumer processes old data with a new understanding of what data belongs to a field number.
```
Comment (shs96c) [inline _posts/2019-01-10-improving-stream-data-quality-with-protobuf-schema-validation.md]: Surely the integer needs to remain stable not over the message's lifespan, but of the message definition's lifespan?

File: _posts/2019-01-10-improving-stream-data-quality-with-protobuf-schema-validation.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+### Backwards Compatibility
+Backwards compatibility means that consumers using a newer version of the schema can read the data produced by a client with an earlier version of the schema. Fields that have been added in the newer version will again be deserialized with default values. Fields that have been deleted in the new schema will not be deserialized, but naturally any subsequent code that was in place to handle that data will need to be refactored to cope.
+
+The remaining Protobuf requirements that are mandated to ensure data consistency are met by ensuring that the ordinal placeholders for each attribute are held immutable throughout a message’s lifespan. In Protobuf the integer field number that represents a field is sacred, as this is what is actually transmitted in a serialized message (as opposed to the field name). All producers and consumers rely on this integer having a consistent meaning, and altering it can cause havoc if a consumer processes old data with a new understanding of what data belongs to a field number. 
+
+The tests we’ve implemented to cover all of these aspects are as follows:
+- Fields numbers must not be amended.
```
Comment (shs96c) [inline _posts/2019-01-10-improving-stream-data-quality-with-protobuf-schema-validation.md]: s/Fields/Field/

File: _posts/2019-01-10-improving-stream-data-quality-with-protobuf-schema-validation.md (hunk 2 above)
Comment (tomseddon) [reply _posts/2019-01-10-improving-stream-data-quality-with-protobuf-schema-validation.md]: @shs96c There was a decision (?) made at deliveroo to adopt American spellings for these types of things. If I were writing anything else I wouldn't dream of using "zee".

I'm going to see if i can dig something up on that decision.

File: _posts/2019-01-10-improving-stream-data-quality-with-protobuf-schema-validation.md [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+
+A key requirement of our centralised message service is resilience and one step towards achieving this is providing 
+guarantees about the structure and data types of messages. These guarantees mean consumer applications can have 
+expectations of the format of the data and be less vulnerable to breaking due to corrupt messages. Another important
+ aspect for resilience is being able to update the data model without breaking consumers that are depending on the 
+ previous version, which means enforcing backwards and forwards compatible schema evolution guarantees. 
+
+Both of these requirements meant that a flexible data format such as JSON would not be ideal and the team invested 
+in researching transport encoding formats the would provide stronger assurances about the data being transmitted. 
+While there are some ways to give greater guarantees for JSON, using for example JSON Schema, this still leaves a
```
Comment (shs96c) [inline _posts/2019-01-10-improving-stream-data-quality-with-protobuf-schema-validation.md]: "using for example" sounds a little clumsy, but it's great to point towards these things. Perhaps link out to https://json-schema.org/

Comment (shs96c) [reply _posts/2019-01-10-improving-stream-data-quality-with-protobuf-schema-validation.md]: "some ways... this" should be "these"

File: _posts/2019-01-10-improving-stream-data-quality-with-protobuf-schema-validation.md [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+guarantees about the structure and data types of messages. These guarantees mean consumer applications can have 
+expectations of the format of the data and be less vulnerable to breaking due to corrupt messages. Another important
+ aspect for resilience is being able to update the data model without breaking consumers that are depending on the 
+ previous version, which means enforcing backwards and forwards compatible schema evolution guarantees. 
+
+Both of these requirements meant that a flexible data format such as JSON would not be ideal and the team invested 
+in researching transport encoding formats the would provide stronger assurances about the data being transmitted. 
+While there are some ways to give greater guarantees for JSON, using for example JSON Schema, this still leaves a 
+lot to be desired, including a lack of well defined mechanisms for schema evolution, and still leaves us with 
+sub-par encoding and decoding performance of JSON itself.
```
Comment (shs96c) [inline _posts/2019-01-10-improving-stream-data-quality-with-protobuf-schema-validation.md]: Is the encoding/decoding speed a limiting factor for us? How does it compare with Node's native marshalling speed for JSON? Do we ever pass protobufs to the front end?

File: _posts/2019-01-10-improving-stream-data-quality-with-protobuf-schema-validation.md [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+ different to that of Protobuf, as it is typically used with a schema definition provided in a header to a file. 
+ Confluent’s schema registry removes this requirement by keeping the schema definition in an API and tagging each 
+ message with a lookup to find that schema. One of the other appealing aspects of Avro is that it manages schema 
+ evolution and backwards and forwards compatibility for you, by keeping track of a writers and a readers schema. 
+
+In the end Avro was discounted as not ideal for Deliveroo’s setup due to lack of cross language support. The thinking
+ behind this was based on a desire for support of generated schema classes in each of Deliveroo’s main supported 
+ languages (Java/Scala/Kotlin, Go and Ruby). Avro only supported the JVM languages in this regard. As it turns out, 
+ the way Avro _does_ support languages outside those with code generation support (through dynamic access to a schema
+  as part of global configuration) turned out to be a feature we also wanted to support with Protobuf.
```
Comment (shs96c) [inline _posts/2019-01-10-improving-stream-data-quality-with-protobuf-schema-validation.md]: Link? The dynamic access to a schema sounds interesting.

File: _posts/2019-01-10-improving-stream-data-quality-with-protobuf-schema-validation.md [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+ all the applications that need to work on that data. In addition to this, benefits such as binary serialisation 
+ (reduced payload size) and schema evolution mechanisms were aspects the team had worked with before on previous 
+ projects, and were keen to make use of again.
+
+We quickly narrowed the choice of serialisation formats to three: Thrift, Protobuf, and Avro. We then proceeded to 
+conduct an evaluation of these formats to determine what would work best for transmission of data over Kafka.
+
+Thrift and Protobuf have very similar semantics, with IDLs that support the broad types and data structures utilised 
+in mainstream programming languages. When conducting our evaluation, we initially chose Thrift due to familiarity, 
+but in the end discounted this due to lack of community support. Avro was an intriguing option, particularly because
```
Comment (joelvanvelden) [inline _posts/2019-01-10-improving-stream-data-quality-with-protobuf-schema-validation.md]: >lack of community support
It was unclear to me if this meant adoption of thrift within the tools we already use (upstream community), or support within the organisation (internal community).

Comment (tomseddon) [reply _posts/2019-01-10-improving-stream-data-quality-with-protobuf-schema-validation.md]: Thanks and yeah you're right. Will clarify.

File: _posts/2019-01-10-improving-stream-data-quality-with-protobuf-schema-validation.md [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+
+## The Need for a Structured Message Format
+
+A key requirement of our centralised message service is resilience and one step towards achieving this is providing 
+guarantees about the structure and data types of messages. These guarantees mean consumer applications can have 
+expectations of the format of the data and be less vulnerable to breaking due to corrupt messages. Another important
+ aspect for resilience is being able to update the data model without breaking consumers that are depending on the 
+ previous version, which means enforcing backwards and forwards compatible schema evolution guarantees. 
+
+Both of these requirements meant that a flexible data format such as JSON would not be ideal and the team invested
```
Comment (mike-ruane) [inline _posts/2019-01-10-improving-stream-data-quality-with-protobuf-schema-validation.md]: I'd be tempted to drop this paragraph completely. Just go straight onto the `Deciding on an Encoding Format` bit. The only thing I can think of going nicely here is a note on the current state; i.e. the use of JSON as a data format at present.

File: _posts/2019-01-10-improving-stream-data-quality-with-protobuf-schema-validation.md [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+ flexibility though, we've implemented both code artefacts for the main languages and a centralised repository for 
+ dynamic access.
+
+## Providing Guarantees on Graceful Schema Evolution
+With our decision on Protobuf confirmed, we turned out attention to creating some extra safeguards around schema 
+evolution.
+
+Some aspects of graceful schema evolution are supported by virtue of Protobuf design. In particular, proto3 has 
+done away with the concept of required fields (which made the decision not to use proto2 easier). With every field 
+being optional, we're already a long way into achieving backwards and forwards compatibility. There were however some
```
Comment (mike-ruane) [inline _posts/2019-01-10-improving-stream-data-quality-with-protobuf-schema-validation.md]: "There were, however, some..."

File: _posts/2019-01-10-improving-stream-data-quality-with-protobuf-schema-validation.md [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+done away with the concept of required fields (which made the decision not to use proto2 easier). With every field 
+being optional, we're already a long way into achieving backwards and forwards compatibility. There were however some 
+other aspects of the format that needed to be enforced in some way.
+
+The Protobuf documentation outlines the [rules for updating messages](https://developers.google
+.com/protocol-buffers/docs/proto3#updating). A summary of those concepts in relation to stream producers and consumers 
+follows.
+
+### Protobuf Properties That Support Forwards Compatibility
+Forwards compatibility means that consumers can read data produced from a client which is using a later version of
```
Comment (mike-ruane) [inline _posts/2019-01-10-improving-stream-data-quality-with-protobuf-schema-validation.md]: "form a client ~which is~ using a ..."

File: _posts/2019-01-10-improving-stream-data-quality-with-protobuf-schema-validation.md [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+ behind this was based on a desire for support of generated schema classes in each of Deliveroo’s main supported 
+ languages (Java/Scala/Kotlin, Go and Ruby). Avro only supported the JVM languages in this regard. As it turns out, 
+ the way [Confluent Schema Registry and Avro](https://docs.confluent.io/current/schema-registry/docs/index.html) 
+ _do_ support languages outside those with code generation support (through dynamic access to a schema as part of 
+ global configuration) turned out to be a feature we also wanted to support with Protobuf. To maintain maximum 
+ flexibility though, we've implemented both code artefacts for the main languages and a centralised repository for 
+ dynamic access.
+
+## Providing Guarantees on Graceful Schema Evolution
+With our decision on Protobuf confirmed, we turned out attention to creating some extra safeguards around schema
```
Comment (krissy) [inline _posts/2019-01-10-improving-stream-data-quality-with-protobuf-schema-validation.md]: Tiny typo: out attention -> our attention

File: _posts/2019-01-10-improving-stream-data-quality-with-protobuf-schema-validation.md [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+## Providing Guarantees on Graceful Schema Evolution
+With our decision on Protobuf confirmed, we turned out attention to creating some extra safeguards around schema 
+evolution.
+
+Some aspects of graceful schema evolution are supported by virtue of Protobuf design. In particular, proto3 has 
+done away with the concept of required fields (which made the decision not to use proto2 easier). With every field 
+being optional, we're already a long way into achieving backwards and forwards compatibility. There were however some 
+other aspects of the format that needed to be enforced in some way.
+
+The Protobuf documentation outlines the [rules for updating messages](https://developers.google
```
Comment (krissy) [inline _posts/2019-01-10-improving-stream-data-quality-with-protobuf-schema-validation.md]: Link isn't formatting due to a space `google .com`

File: _posts/2019-01-10-improving-stream-data-quality-with-protobuf-schema-validation.md [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+
+If a client serialises a message with a missing topic definition or mismatched definition in relation to the topic 
+being published to, the Producer API returns a 400 Bad Request to that client.
+
+The full contract is achieved through a combination of the topic metadata tying a topic to a schema and a separate 
+mapping between an authentication key used by publishers which maps to one and only one topic. By ensuring all 
+publishing to Kafka is done via our Stream Producer API (topic ACLs prevent any other  applications from publishing)
+, we have implemented a method to enforce the relationship between producers, topics and schemas.
+
+## Managing Schema Artefacts and the Path Towards a Dynamic Registry
```
Comment (krissy) [inline _posts/2019-01-10-improving-stream-data-quality-with-protobuf-schema-validation.md]: Suggest breaking this para up in 2 or 3 as it has some key points but they get a little lost in the length (as well as being at the end of the post :))

**Pull request merged**